packages/typegpu/src/core/function/tgpuFn.ts

```diff
@@ -0,0 +1,172 @@
+import { inGPUMode } from '../../gpuMode';
+import type { TgpuNamable } from '../../namable';
+import { valueList } from '../../resolutionUtils';
+import type { Block } from '../../smol';
+import { code } from '../../tgpuCode';
+import type {
+  AnyTgpuData,
+  ResolutionCtx,
+  TgpuResolvable,
+  Wgsl,
+} from '../../types';
+import { createFnCore } from './fnCore';
+import type { Implementation, UnwrapArgs, UnwrapReturn } from './fnTypes';
+
+// ----------
+// Public API
+// ----------
+
+/**
+ * Describes a function signature (its arguments and return type)
+ */
+export interface TgpuFnShell<
+  Args extends AnyTgpuData[],
+  Return extends AnyTgpuData | undefined,
+> {
+  readonly argTypes: Args;
+  readonly returnType: Return | undefined;
+
+  /**
+   * Creates a type-safe implementation of this signature
+   */
+  implement(
+    implementation: (...args: UnwrapArgs<Args>) => UnwrapReturn<Return>,
+  ): TgpuFn<Args, Return>;
+
+  /**
+   * @param implementation
+   *   Raw WGSL function implementation with header and body
+   *   without `fn` keyword and function name
+   *   e.g. `"(x: f32) -> f32 { return x; }"`;
+   */
+  implement(implementation: string): TgpuFn<Args, Return>;
+}
+
+interface TgpuFnBase<
+  Args extends AnyTgpuData[],
+  Return extends AnyTgpuData | undefined = undefined,
+> extends TgpuResolvable,
+    TgpuNamable {
+  readonly shell: TgpuFnShell<Args, Return>;
+
+  $uses(dependencyMap: Record<string, unknown>): this;
+  $__ast(argNames: string[], body: Block): this;
+}
+
+export type TgpuFn<
+  Args extends AnyTgpuData[],
+  Return extends AnyTgpuData | undefined = undefined,
+> = TgpuFnBase<Args, Return> &
+  ((...args: UnwrapArgs<Args>) => UnwrapReturn<Return>);
+
+export function fn<Args extends [...AnyTgpuData[]] | []>(
```

Comment:
curious what's the reasoning behind `[...AnyTgpuData[]]` vs `AnyTgpuData[]`

Reply:
I could've sworn it did not work without it, haha. Seems like `AnyTgpuData[] | []` still behaves correctly, but just `AnyTgpuData[]` starts to break down.

Reply:
Without it, the type of function arguments is inferred to be an array of a union, not a tuple, so for example `(F32|U32)[]` instead of `[F32, U32]`